tokio/src/io/util/copy.rs

```diff
@@ -82,6 +83,8 @@ impl CopyBuffer {
         R: AsyncRead + ?Sized,
         W: AsyncWrite + ?Sized,
     {
+        ready!(crate::trace::trace_leaf(cx));
+        ready!(poll_proceed_and_make_progress(cx));
```

Comment:
You should only make progress if it actually copied data.

See the mpsc channel for an example:

https://github.com/tokio-rs/tokio/blob/7c606ab44aa9f0c33dfcf5bc8678411dddeaa9e5/tokio/src/sync/mpsc/chan.rs#L250-L251

https://github.com/tokio-rs/tokio/blob/7c606ab44aa9f0c33dfcf5bc8678411dddeaa9e5/tokio/src/sync/mpsc/chan.rs#L259-L263

Reply:
Thank you for your review.
How about now? After filling, writing, and flushing data, I called `make_progress`.

Reply:
Like in the examples I linked, you need to call `crate::runtime::coop::poll_proceed` at the start of the function, and then call `coop.made_progress()` in the places where you are currently calling `poll_proceed_and_make_progress`.

It's not enough to just call it when it exits. The `poll_proceed_and_make_progress` function can only be used for functions that never return `Poll::Pending`.

Reply:
> Like in the examples I linked, you need to call `crate::runtime::coop::poll_proceed` at the start of the function, and then call `coop.made_progress()`

But the crate::runtime::coop::poll_proceed only exists if we enable the coop feature. What should we do if we don't have the `coop` feature?

Reply:
Then do nothing.

Reply:
Updated. I just used `#[cfg(any(
                    feature = "fs",
                    feature = "io-std",
                    feature = "net",
                    feature = "process",
                    feature = "rt",
                    feature = "signal",
                    feature = "sync",
                    feature = "time",
                    ))]` to control it. I am not sure if we have a better way to do it. Thanks for your help. Sorry for those stupid questions. 😢

Reply:
I believe we have [a macro](https://github.com/tokio-rs/tokio/blob/2f730d4d5a4810088428f6e6397bda42d376deec/tokio/src/macros/cfg.rs#L487-L503) for doing that. You should be able to write
```rs
cfg_coop! {
    let coop = ready!(crate::runtime::coop::poll_proceed(cx));
}
```
And similar for the calls to `make_progress`. The macro does not actually introduce a scope, so I believe this should work.

Your questions are not stupid :). You're asking questions about internal Tokio APIs that don't exist elsewhere. It's no surprise that you are not familiar with them. If we did not have the `cfg_coop!` macro, then what you did would be the only way to do it.

Reply:
Thanks for your reply.

Actually, I tried it. But it seemed this macro only supports starting with an item keyword. 

Do you think I should extend it to support it? (I will try, but I am unsure if it is doable).